Frameworks/Foundation/NSDecimal.mm

```diff
@@ -0,0 +1,592 @@
+
+    if (NEGATIVE_FLAG(leftOperand) != NEGATIVE_FLAG(rightOperand)) {
+        // miss match in sign.
+        return IS_NEGATIVE(rightOperand) ? NSOrderedDescending : NSOrderedAscending;
+    }
+
+    int leftValue = EXPONENT(leftOperand) + LENGTH(leftOperand);
+    int rightValue = EXPONENT(rightOperand) + LENGTH(rightOperand);
+    if (leftValue > rightValue) {
+        return IS_NEGATIVE(rightOperand) ? NSOrderedAscending : NSOrderedDescending;
+    }
+    if (leftValue < rightValue) {
+        return IS_NEGATIVE(rightOperand) ? NSOrderedDescending : NSOrderedAscending;
+    }
+
+    // same length
+    int index = std::min(LENGTH(leftOperand), LENGTH(rightOperand)) - 1;
+    while (index >= 0) {
+        if (MANTISSA(leftOperand)[index] < MANTISSA(rightOperand)[index]) {
+            return (IS_NEGATIVE(rightOperand)) ? NSOrderedDescending : NSOrderedAscending;
+        }
+
+        if (MANTISSA(leftOperand)[index] > MANTISSA(rightOperand)[index]) {
+            return IS_NEGATIVE(rightOperand) ? NSOrderedAscending : NSOrderedDescending;
+        }
+        index--;
+    }
+
+    if (LENGTH(rightOperand) > LENGTH(leftOperand)) {
+        return IS_NEGATIVE(rightOperand) ? NSOrderedDescending : NSOrderedAscending;
+    }
+    if (LENGTH(rightOperand) < LENGTH(leftOperand)) {
+        return IS_NEGATIVE(rightOperand) ? NSOrderedAscending : NSOrderedDescending;
+    }
+
+    return NSOrderedSame;
+}
+
+/**
+ @Status Interoperable
+*/
+void NSDecimalCopy(NSDecimal* destination, const NSDecimal* source) {
+    if (!source || !destination) {
+        return;
+    }
+
+    NEGATIVE_FLAG(destination) = NEGATIVE_FLAG(source);
+    EXPONENT(destination) = EXPONENT(source);
+    LENGTH(destination) = LENGTH(source);
+    destination->_isCompact = source->_isCompact;
+    destination->_reserved = source->_reserved;
+    memcpy(&MANTISSA(destination), &MANTISSA(source), NSDecimalMaxSize * sizeof(unsigned short));
+}
+
+/**
+ @Status Stub
+ @Notes
+*/
+NSCalculationError NSDecimalDivide(NSDecimal* result,
+                                   const NSDecimal* leftOperand,
+                                   const NSDecimal* rightOperand,
+                                   NSRoundingMode roundingMode) {
+    UNIMPLEMENTED();
+    return StubReturn();
+}
+
+/**
+ @Status Stub
+ @Notes
+*/
+NSCalculationError NSDecimalMultiply(NSDecimal* result,
+                                     const NSDecimal* leftOperand,
+                                     const NSDecimal* rightOperand,
+                                     NSRoundingMode roundingMode) {
+    UNIMPLEMENTED();
+    return StubReturn();
+}
+
+/**
+ @Status Interoperable
+*/
+NSCalculationError NSDecimalMultiplyByPowerOf10(NSDecimal* result, const NSDecimal* number, short power, NSRoundingMode roundingMode) {
+    THROW_NS_IF_FALSE(E_INVALIDARG, (result != NULL) || (number != NULL));
+
+    NSDecimalCopy(result, number);
+    int exponent;
```

Comment:
>exponent [](start = 8, length = 8)

nit: initialize exponent = EXPONENT(result) + power here instead of in the if statement.